### ISSUE TITLE
Footnotes in subfigure captions

### ISSUE THREAD

**Issue by esnahn** [lierdakil/pandoc-crossref#292]
Problem: With styles like [Chicago](https://csl.mendeley.com/styleInfo/?styleId=http%3A%2F%2Fwww.zotero.org%2Fstyles%2Fchicago-fullnote-bibliography) that uses footnotes for in-text citations, Usage of references in caption of subfigure divs would break PDF creation.

Solution: Adding `\protect`? Fixing citeproc's behavior in `\caption`?

Source:
```markdown
<div id="fig:sub">

![a](test.jpg){}
![b](test.jpg){}

sub [@test]
</div>

![b[@test]](test.jpg){}
```

TeX translation:
```latex
\caption{sub\footnote{T. Test, {``{On testing footnotes in subfigure
  captions},''} 2021.}}
...
\caption[b]{b\footnotemark{}}
\end{figure}
\footnotetext{Test.}
```

As you can see, references in subfigure divs don't get the same treatment as normal figures (`\footnote` vs `\footnotemark` and short caption).

If I try to produce pdf, I get this:
```console
$ pandoc test.md -o test.pdf --bibliography=test.bib --csl=chicago-fullnote-bibliography.csl --pdf-engine=xelatex --filter pandoc-crossref --citeproc
Error producing PDF.
! Argument of \caption@ydblarg has an extra }.
<inserted text>
                \par
l.114   captions},''} 2021.}}
```

If I add `\protect` before `\footnote`, I can get it work; but I am not sure that's the right way to do...

```latex
\caption{sub\protect\footnote{T. Test, {``{On testing footnotes in subfigure
  captions},''} 2021.}}
```

```console
$ pandoc test-fix.md -o test.pdf --bibliography=test.bib --csl=chicago-fullnote-bibliography.csl --pdf-engine=xelatex --filter pandoc-crossref --citeproc
```

- pandoc.exe 2.11.2
- pandoc-crossref v0.3.8.4 git commit ec2d4bc2eeaac30686147d0549abdd6e4af0776d (HEAD) built with Pandoc v2.11.2, pandoc-types v1.22 and GHC 8.10.2

[footnote.zip](https://github.com/lierdakil/pandoc-crossref/files/5794203/footnote.zip)

**Comment by lierdakil** [comment]
Hmm. This is actually pretty perplexing.

Pandoc doesn't do subfigures, so I have to piece that together by hand from raw blocks and inlines. At the same time, I can't render the caption in-place (that would lose all context), so I have to pass it through. But Pandoc changes LaTeX output behaviour slightly depending on whether something is "inside" a caption or not. As far as I can tell, there is no way to tell Pandoc it should treat some fragment as being "inside" a caption.

@jgm sorry for the ping, but any ideas? I don't seem to have any. Well, save for adding subfigures directly to LaTeX writer (and perhaps reader). For reference, here is the relevant code snippet:
https://github.com/lierdakil/pandoc-crossref/blob/753142fc7a0e0be648a7606d425d01819db7aaf8/lib/Text/Pandoc/CrossRef/References/Blocks.hs#L141-L147 `cont` also can contain captions, hence footnotes could also appear there.

I guess I could `minipage` the whole thing and `protect` the notes, but that's not necessarily the expected formatting, since notes would appear at the bottom of the figure, not at the bottom of the page. Simply `protect`ing the notes is pretty unreliable as far as I know.

**Comment by jgm** [comment]
Sorry, no ideas here other than the minipage suggestion.

**Comment by lierdakil** [comment]
@esnahn, would formatting like this be acceptable?

![A picture showing footnote text immediately after the figure caption](https://user-images.githubusercontent.com/7275622/104303838-a5226d80-54db-11eb-92f5-999df5e0d560.png)

Footnotes in floats can be pretty wonky anyway (with footnotes appearing before floats themselves), so formatting like this is a bit more consistent if nothing else. I appreciate this might not be exactly what you're looking for, but as I said above, I'm pretty short on ideas how to make it work otherwise without adding very specific hacks into pandoc itself (which I'd prefer to avoid)

**Comment by esnahn** [comment]
Anything is acceptable when the other option is failing to produce PDF, actually. As for me, I will use `\protect` hack for this time.

**Comment by lierdakil** [comment]
@esnahn, I feel like I need to clarify: `\protect` by itself will more often than not lead to issues, particularly missing footnotes. For example: [test.pdf](https://github.com/lierdakil/pandoc-crossref/files/5801711/test.pdf) -- notice how only footnote 3 is rendered, and 1 and 2 are referenced, but are in fact missing.

**Comment by esnahn** [comment]
Thanks for the warning. With that example, it looks clear that protect is not a good (sound) solution...

**Comment by lierdakil** [comment]
I think I figured out a way to do this more or less in-line with Pandoc, but it requires some pretty arcane LaTeX wizardry. Not entirely sure it's worth it to be perfectly honest, but I've pushed a proof of concept. Binaries should be available on [bintray](https://bintray.com/lierdakil/pandoc-crossref/master/c2605ce12fb64bd9d7b0830a95cd1bdbde05c413) once CI finishes (in about an hour, provided nothing breaks)

**Comment by lierdakil** [comment]
@esnahn, binaries [are up](https://bintray.com/lierdakil/pandoc-crossref/master/c2605ce12fb64bd9d7b0830a95cd1bdbde05c413#files). Would be great if you could test those for your actual use case and report if there are any issues. Thanks in advance.

**Comment by esnahn** [comment]
@lierdakil, A sorcery indeed! I can't understand all of it, but I believe that `\footnote` works differently in `pandoccrossrefsubfigures` env., and dropping all of `\footnotetext` at the end, right? I was doing same thing manually, so thank you!

It leaves a small inconsistency, still. Pandoc removes footnotes from list of figures, by using short notes in `\caption`.
```
...
\caption[b]{b\footnotemark{}}
...
```

I believe as long as crossref runs before pandoc does, crossref can't fix pandoc failing to recognize it being inside `\caption`, correct? I can still use pandoccrossrefsubfigures env. and fix one line, and it is much better result than my own hack.

**Comment by lierdakil** [comment]
> can crossref recognize citeproc tags and drop them in the short captions?

It can do that indeed. There's a problem however: pandoc will only remove citations from the short caption when those are rendered as footnotes (actually, pandoc will only remove footnotes, period -- it just does that after citeproc runs). And pandoc-crossref runs before citation processing takes place (by design) so it's not obvious if those should be removed or not.

I can suggest a workaround -- run the following Lua filter after citeproc (i.e. `--citeproc -L shortcaptions.lua` if you name the file `shortcaptions.lua`)

```lua
function Inlines(ils)
  i = 1
  while ils[i] do
    if ils[i].t == "RawInline" and ils[i].format == "latex" and ils[i].text == "\\caption" then
      if ils[i+1] and ils[i+1].t == "Span" then
        ils[i].text = ils[i].text .. "["
        ils:insert(i+1, pandoc.RawInline("latex", "]"))
        ils:insert(i+1, pandoc.walk_inline(ils[i+2], {
          Inline = function(elem)
            if elem.t == "Note" then
              return pandoc.Str("")
            else
              return elem
            end
          end
        }))
      end
    end
    i = i + 1
  end
  return ils
end
```
It's a bit of a dirty hack, but, well, it should work, which is better than nothing.

**Comment by lierdakil** [comment]
As a heads-up, and also for posterity, I will be releasing v2.9.1 shortly which generates "short" captions for subfigures internally with footnotes filtered out. This still doesn't handle the issue of citations being rendered as footnotes, so post-processing is still required. However, the lua filter above needs to be changed slightly:

```lua
function Inlines(ils)
  i = 1
  while ils[i] do
    if ils[i].t == "RawInline" and ils[i].format == "latex" and ils[i].text == "\\caption[" then
      if ils[i+1] and ils[i+1].t == "Span" then
        ils[i+1] = pandoc.walk_inline(ils[i+1], {
          Inline = function(elem)
            if elem.t == "Note" then
              return pandoc.Str("")
            end
          end
        })
      end
    end
    i = i + 1
  end
  return ils
end
```

Note, however, that this filter will just remove notes, so it needs to be run after citeproc, and it may mess up how references are presented. If you know that citations are rendered as footnotes, it might be a better option to remove those from short captions by running the following filter **before** citeproc, but **after** crossref:

```lua
function Inlines(ils)
  i = 1
  while ils[i] do
    if ils[i].t == "RawInline" and ils[i].format == "latex" and ils[i].text == "\\caption[" then
      if ils[i+1] and ils[i+1].t == "Span" then
        ils[i+1] = pandoc.walk_inline(ils[i+1], {
          Inline = function(elem)
            if elem.t == "Cite" then
              return pandoc.Str("")
            end
          end
        })
      end
    end
    i = i + 1
  end
  return ils
end
```

TL;DR, my recommendation is to use the second filter _after_ pandoc-crossref but _before_ citeproc if you know citations will be rendered as footnotes.

This is only relevant for LaTeX output.

**Comment by esnahn** [comment]
> If you know that citations are rendered as footnotes

That is defined in csl file, with `class="note"`. Could crossref reads the csl option/file and decides it for me?

Also, if it is to run before citeproc, maybe someday, could it be internalized in crossref?

**Comment by lierdakil** [comment]
> Could crossref reads the csl option/file and decides it for me?

Parsing csl definitions is wildly out of scope.

> Also, if it is to run before citeproc, maybe someday, could it be internalized in crossref?

The of a separate filter is this: you don't _necessarily_ want to remove citations from short captions (indeed, in some conventions it's undesirable to remove those). Since there's no way for pandoc-crossref to know your intent, I'm opting to provide a reasonable workaround instead of trying to build a falling tower of brittle heuristics.
